apps/astarte_appengine_api/lib/astarte_appengine_api/device/queries.ex

```diff
@@ -574,30 +593,24 @@ defmodule Astarte.AppEngine.API.Device.Queries do
     end
   end
 
-  def delete_attribute(realm_name, client, device_id, attribute_key) do
+  def delete_attribute(realm_name, _client, device_id, attribute_key) do
     keyspace = keyspace_name(realm_name)
     query = from(d in DatabaseDevice, select: d.attributes)
     opts = [prefix: keyspace, consistency: :quorum]
 
     with {:ok, attributes} <- Repo.fetch(query, device_id, opts),
          {:ok, _} <- get_value(attributes, attribute_key, :attribute_key_not_found) do
-      delete_attribute_statement = """
-        DELETE attributes[:attribute_key]
-        FROM devices
-        WHERE device_id = :device_id
-      """
-
-      delete_attribute_query =
-        DatabaseQuery.new()
-        |> DatabaseQuery.statement(delete_attribute_statement)
-        |> DatabaseQuery.put(:attribute_key, attribute_key)
-        |> DatabaseQuery.put(:device_id, device_id)
-        |> DatabaseQuery.consistency(:each_quorum)
-
-      case DatabaseQuery.call(client, delete_attribute_query) do
-        {:ok, _result} ->
-          :ok
+      map_new_attribute = MapSet.new([attribute_key])
 
+      query_delete_attributes =
+        from ApiDevice,
+          prefix: ^keyspace,
+          where: [device_id: ^device_id],
+          update: [set: [attributes: fragment("attributes - ?", ^map_new_attribute)]]
+
+      with {1, nil} <- Repo.update_all(query_delete_attributes, [], consistency: :each_quorum) do
+        :ok
+      else
```

Comment:
We're already verifying that the attribute exists with the first query above.
Here we can assert that the call to `Repo.update_all` must edit 1 row.
Also note that Repo.update_all always returns a `{count, rows}` tuple, so we don't need to handle other type of return values as before.

The whole with/else block can become:

```elixir
{1, nil} = Repo.update_all(query_delete_attributes, [], consistency: :each_quorum)

:ok
```